framework/src/main/java/org/checkerframework/common/wholeprograminference/SceneToStubWriter.java

```diff
@@ -0,0 +1,563 @@
+                    throw new AssertionError();
+                }
+            } else {
+
+                for (Object o2 : l) {
+                    sj.add(formatAnnotationValue(aaft.elementType, o2));
+                }
+            }
+            return sj.toString();
+        } else if (aft instanceof ClassTokenAFT) {
+            return aft.format(o);
+        } else if (aft instanceof BasicAFT && o instanceof String) {
+            return Strings.escape((String) o);
+        } else if (aft instanceof BasicAFT && o instanceof Long) {
+            return o.toString() + "L";
+        } else {
+            return o.toString();
+        }
+    }
+
+    /**
+     * Returns the String representation of an annotation in Java source format.
+     *
+     * @param a the annotation to print
+     * @return the formatted annotation
+     */
+    private static String formatAnnotation(Annotation a) {
+        String annoName = a.def().name.substring(a.def().name.lastIndexOf('.') + 1);
+        if (a.fieldValues.isEmpty()) {
+            return "@" + annoName;
+        }
+        StringJoiner sj = new StringJoiner(",", "@" + annoName + "(", ")");
+        for (Map.Entry<String, Object> f : a.fieldValues.entrySet()) {
+            AnnotationFieldType aft = a.def().fieldTypes.get(f.getKey());
+            sj.add(f.getKey() + "=" + formatAnnotationValue(aft, f.getValue()));
+        }
+        return sj.toString();
+    }
+
+    /**
+     * Returns all annotations in {@code annos} in a form suitable to be printed as Java source
+     * code.
+     *
+     * <p>Each annotation is followed by a space, to separate it from following Java code.
+     *
+     * @param annos the annotations to format
+     * @return all annotations in {@code annos}, separated by spaces, in a form suitable to be
+     *     printed as Java source code
+     */
+    private static String formatAnnotations(Collection<? extends Annotation> annos) {
+        StringBuilder sb = new StringBuilder();
+        for (Annotation tla : annos) {
+            if (!isInternalJDKAnnotation(tla.def.name)) {
+                sb.append(formatAnnotation(tla));
+                sb.append(" ");
+            }
+        }
+        return sb.toString();
+    }
+
+    /**
+     * Formats the component types of an array via recursive descent through the array's scene-lib
+     * structure.
+     *
+     * @param e the array's scenelib type element
+     * @param arrayType the string representation of the array's type
+     * @return the type formatted to be written to Java source code
+     */
+    private static String formatArrayType(ATypeElement e, String arrayType) {
+        StringBuilder result = new StringBuilder();
+        return formatArrayTypeImpl(e, arrayType, result);
+    }
+
+    /**
+     * The implementation of formatArrayType. Because of the (insane) structure of scenelib, the
+     * descent order is innermost array type to outermost array type, then component type. So if we
```

Comment:
I'm not sure that the example on the next line is an example of innermost to outermost.
In your example, I think `@Bar` then `@Baz` then `@Foo` is outermost to innermost.
See https://checkerframework.org/jsr308/specification/java-annotation-design.html#array-syntax for an explanation of array syntax.

I'm not questioning your code at this point, just trying to understand the document.